plugin/utils/symbol_table.py

```diff
@@ -0,0 +1,109 @@
+# Copyright 2019 Google LLC
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+#         * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#         * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#         * Neither the name of Google Inc. nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+A utility class used to get and store unique symbols.
+"""
+
+import copy
+
+
+class SymbolTable(object):
+
+    """
+    Initialize a case-sensitive SymbolTable with a set of symbols.
+    """
+    def __init__(self):
+        self.symbol_table = copy.deepcopy(java_reserved_symbols)
+
+    def getNewSymbol(self, desired_name):
+
+        if not (desired_name in self.symbol_table):
+            self.symbol_table.add(desired_name)
+            return desired_name
+
+        # Resolve collisions with one or more underscores.
+        while (desired_name + "_") in self.symbol_table:
+            self.symbol_table.add(desired_name + "_")
+        return desired_name + "_"
+
+
+java_reserved_symbols = {
+    "abstract",
```

Comment:
This set is ripped from https://github.com/googleapis/gapic-generator/blob/8a4ad18ccd5eb3869da7e95cdf7ed3c76ea94281/src/main/java/com/google/api/codegen/util/java/JavaNameFormatter.java#L107